watcher/watcher.go

```diff
@@ -25,35 +25,90 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+type BurnLogResult struct {
+	Result ethereumbindings.MintGatewayLogicV1LogBurn
+	Error  error
+}
+
+type BurnLogFetcher interface {
+	FetchBurnLogs(ctx context.Context, from uint64, to uint64) (chan BurnLogResult, error)
+}
+
+type EthBurnLogFetcher struct {
+	bindings *ethereumbindings.MintGatewayLogicV1
+}
+
+func NewBurnLogFetcher(bindings *ethereumbindings.MintGatewayLogicV1) EthBurnLogFetcher {
+	return EthBurnLogFetcher{
+		bindings: bindings,
+	}
+}
+
+// This will fetch the burn event logs using the ethereum bindings and emit them via a channel
+// We do this so that we can unit test the log handling without calling ethereum
+func (fetcher EthBurnLogFetcher) FetchBurnLogs(ctx context.Context, from uint64, to uint64) (chan BurnLogResult, error) {
+	iter, err := fetcher.bindings.FilterLogBurn(
+		&bind.FilterOpts{
+			Context: ctx,
+			Start:   from,
+			End:     &to,
+		},
+		nil,
+		nil,
+	)
+	if err != nil {
+		return nil, err
+	}
+	resultChan := make(chan BurnLogResult)
+
+	go func() {
+		for iter.Next() {
+			err := iter.Error()
```

Comment:
I think the error should be checked outside the `iter.Next()` 
```golang
for iter.Next(){
      // ...
}
if iter.Error() != nil {
     // ...
}

```
Also I notice there's `iter.Close()` function we can use. Might not be the reason causing the issue, but could release some resources as a small optimization.

Reply:
Doesn't the iter close when it's done anyhow?

Reply:
Doesn't look like it. The binding creates a `event.Subscription` when initializing the iterator. I can't see it get `Unsubscribe` anywhere expect the `iter.Close()` function. I think we can do something like
```golang
iter, err := bindings.FilterLogBurn()
if err != nil {
    // ...
}
defer iter.Close()
```
Event the subscription gets unsubscribed elsewhere, it should be fine as the actual implementation uses `sync.Once` to ignore extra calls.

Reply:
Ok, addressed this